table.go

```diff
@@ -51,64 +52,144 @@ type RowDiff struct {
 	Changes map[string]*ValueChange `json:"changes"`
 }
 
-// Diff takes two tables and diffs them. If diffRows is true, value-level changes
-// will be reported as well.
-func Diff(t1, t2 Table, diffRows bool) (*TableDiff, error) {
+const (
+	EventColumnAdded   = "column-added"
+	EventColumnChanged = "column-changed"
+	EventColumnRemoved = "column-removed"
+	EventRowAdded      = "row-added"
+	EventRowChanged    = "row-changed"
+	EventRowRemoved    = "row-removed"
+)
+
+type Event struct {
+	Type    string                  `json:"type"`
+	Offset  int64                   `json:"offset,omitempty"`
+	Column  string                  `json:"column,omitempty"`
+	OldType string                  `json:"old_type,omitempty"`
+	NewType string                  `json:"new_type,omitempty"`
+	Key     map[string]interface{}  `json:"key,omitempty"`
+	Data    map[string]interface{}  `json:"data,omitempty"`
+	Changes map[string]*ValueChange `json:"changes,omitempty"`
+}
+
+func compareRows(r1, r2 [][]byte) int {
+	for i, v1 := range r1 {
+		if p := bytes.Compare(v1, r2[i]); p != 0 {
+			return p
+		}
+	}
+	return 0
+}
+
+func newValueMap(r Row, cols map[string]string) map[string]interface{} {
+	c := make(map[string]interface{}, len(cols))
+	for k := range cols {
+		c[k] = r.Value(k)
+	}
+
+	return c
+}
+
+func newKeyMap(r Row, cols []string) map[string]interface{} {
+	c := make(map[string]interface{}, len(cols))
+	for _, k := range cols {
+		c[k] = r.Value(k)
+	}
+
+	return c
+
+}
+
+func DiffEvents(t1, t2 Table, h func(e *Event)) error {
 	key1 := t1.Key()
 	key2 := t2.Key()
 
 	if len(key1) == 0 || len(key2) == 0 {
-		return nil, errors.New("a key must be provided")
+		return errors.New("a key must be provided")
 	}
 
 	if len(key1) != len(key2) {
-		return nil, errors.New("keys are different lengths")
-	}
-
-	diff := TableDiff{
-		ColsAdded:   make([]string, 0),
-		ColsDropped: make([]string, 0),
-		TypeChanges: make(map[string]*TypeChange),
+		return errors.New("keys are different lengths")
 	}
 
 	cols1 := t1.Cols()
 	cols2 := t2.Cols()
 
-	// For lookup.
-	keyMap := make(map[string]struct{}, len(key1))
+	// Build lookups for key columns.
+	// Validate both tables have the key columns.
+	key1Map := make(map[string]struct{}, len(key1))
 	for _, c := range key1 {
-		keyMap[c] = struct{}{}
+		if _, ok := cols1[c]; !ok {
+			return fmt.Errorf("table 1 does not have key column `%s`", c)
+		}
+		key1Map[c] = struct{}{}
+	}
+
+	key2Map := make(map[string]struct{}, len(key2))
+	for _, c := range key2 {
+		if _, ok := cols2[c]; !ok {
+			return fmt.Errorf("table 2 does not have key column `%s`", c)
+		}
+		key2Map[c] = struct{}{}
 	}
 
 	// Columns to check when comparing rows.
-	var cmpcols []string
+	var (
+		cmpCols  []string
+		dropCols []string
+		newCols  []string
+	)
 
 	for c, ty1 := range cols1 {
+		// Ignore key columns.
+		if _, ok := key1Map[c]; ok {
```

Comment:
I think there are still some subtle `key1 != key2` edge cases that might have unexpected results due to this logic. For example, what happens if:

```
cols1 = ['A', 'B', 'C', 'D']
key1 = ['A']
cols2 = ['A', 'B', 'C', 'D', 'A1']
key2 = ['A1']
```

I think that any changes to data in column `'A'` will be silently passed over. A real life scenario where this might occur is if someone realizes that their key column has domain meaning and is non-unique so they copy `'A'` to '`A1`' and make changes to the data in `'A'`, then want to view the changes. Obviously that's not an ideal scenario and some bad decisions have been made, but I wouldn't be too surprised if we ran across it at some point.

It seems to me like all of this logic to partition columns into a non-key comparison set, a new set, and a dropped set could be avoided if we accepted more calls to `bytes.Equal(r1.Bytes(c), r2.Bytes(c))` on line 316. How expensive is that operation? Or, we could add some more logic that includes key columns in the comparison set when `key1 != key2`. Thoughts?

Reply:
This is partly a naming issue. The assumption here is that if you have columns of the same name, they will be compared. Key columns do not need to be compared (at least with one another) when performing the record-level diff since they are equal by definition. The question is whether a key-column on one side with the same name should be compared with the same named column on the other side if it is not a key column (`A` in `cols2` with `A` in `cols1`).

If a user simply changed the column names so non-key columns don't overlap with key columns in the other table, then this wouldn't an issue. There is no column mapping in the API partly because the original use case involved diffing SQL tables with identical schemas. It may be more useful for CSV files where you would rather not change the header.

Reply:
So options would be to 1) document that currently the tool only supports simple adding and dropping of columns (with plans to add column mapping later), or 2) add column mapping now.

@aaron0browne Good observations.

Aside: my only contribution is to say that I had either not seen or had forgotten the Go idiom for sets; wowza. I think I've used a boolean value in the map for that purpose, but the 0-storage `struct{}{}` is, uh, genius.

Reply:
@murphyke I suppose adding column mapping now would be fine. Grant it is technically out of the scope of this PR, but that would be the next thing to do.

Agreed the `map[string]struct{}` "set" structure is wild. I can't take the credit for coming up with the idea, but have used it since I learned about it.

Reply:
So, I don't want to hold this up for too long, but I did want to make my suggestions concrete, so I submitted them in #5.